[TASK] Switch from jQuery to vanilla JavaScript

rintisch opened the pull request:
This renews the PR for the outdated PR #431.

For simplicity I will copy the description from the old PR.

___

This is WIP because some points are not final:

* There are two spots in the JS marked with "Todo" (simply search for `//Todo`)
* The JavaScript for the form finisher is not included anywhere. Therefore it is not tested as well.

This ticket solves https://github.com/extcode/cart/issues/333.

This fixes also https://docs.typo3.org/p/extcode/cart/8.6/en-us/Changelog/8.5/Important-404-ChangeCheckboxValueInOrderForm.html for the multistep checkout where this adaption was not done.

extcode commented:
@rintisch Thank you for your contribution.

You asked:
> // TODO: What's the meaning of that?
> // $(document).trigger("status.cartWasChanged", [true]);

My answer:
The idea behind this trigger was that cart provides an "event" at this point that you can listen to with your own JavaScript in order to be able to place an effect on the shopping cart icon in addition to updating the DOM.

My experience shows that it is always difficult to replace the entire JavaScript if you only want to adjust one small thing. For example, you could listen to the event and add a class to the mini shopping cart in the navigation area of the page.

As you have also split this into smaller files, this should be easier. I don't know how to provide such events with pure JavaScript and whether they could be used elsewhere. For example, when a message is displayed when a product is added. At this point, you could trigger an event and provide a standard event listener, which can then be replaced more easily.

The second ToDo is a little more complicated. This currently only works if you have products that can be customized with a Form Framework Form before the product is placed in the shopping cart. I'll send you an example with extcode/cart-events via Slack.
That would be too specific here in the ticket.

koehnlein commented:
Here my proposal for the vanilla JavaScript event: https://github.com/rintisch/cart/pull/1

* better name with prefix `extcode.cart.`
* enriched with response data

I really used the old event and would miss it, if there would be no event anymore.

extcode commented:
@koehnlein Do you see any other places where it would be helpful to provide an event in JavaScript? I'm not that deep into the frontend topics and don't really know the requirements very well. You are the second person I know of who actually uses this event.

koehnlein commented:
@extcode At the moment I personally do not need further JavaScript events in my project.

But I just checked the JavaScript files in this pull request and maybe useful events could be added:

* in `showMessageBlock()` to override rules for show text message
* in `fadeOut()` to override rules to hide the text message

I'm not familiar what the other JavaScript files do. So I currently cannot give any suggestions for them.

rintisch commented:
@extcode @koehnlein 
I will have a look at the code next week and update the PR!

koehnlein commented:
@rintisch Tiny reminder :wink:

rintisch commented:
@extcode @koehnlein I updated the JS.

My inital trial used ES modules but the TYPO3 assets collector has no option to render `<script type="module" ...>`. I could include it with `additionalAttributes` but I am not aware off all consequences (what happens if `compressJS` is set in TypoScript? - I have not tested it). 
The consequence: now  it isI made it really unDRY with a lot of copy+paste :( I guess we will need to add a build step in a next step.

@extcode I have not tested the code for the formfinisher (which is no included in `add_to_cart.js` because otherwise the long functions for updating the minicart and to render the success message would also be duplicated.)

koehnlein commented:
Sorry, my project went a bit offtopic. :sweat: I had expected, I can test the new JS event implementation today. But looks, like it will last one more week.

koehnlein commented on Resources/Public/JavaScript/add_to_cart.js:
```diff
@@ -0,0 +1,215 @@
+
+    function dispatchCustomEvent(name, dataObject) {
+        const customEvent = new CustomEvent(
+            `extcode:{name}`,
```
Did you mean?

```suggestion
            `extcode:${name}`,
```

Because otherwise, this line simply renders as string `'extcode:{name}'` in all browsers

rintisch replied:
@koehnlein 
Of course did I miss the $ sign (and copied it over and over again...) Thanks for the hint. I force-pushed an update.

koehnlein commented:
@rintisch @extcode JavaScript works fine for me

:+1:

rintisch commented:
@koehnlein 
Thanks for testing! I may need to move some JS from EXT:cart_products to EXT:cart (still need feedback from @extcode (wrote on slack)), afterwards I will switch the mode of the PR from draft to mergeable.

extcode commented:
I will have a look. @rintisch Is it best practice to put _package,json_ and _package-lock.json_ in to the _Resources/Private/JavaScript/_ directory and not into the project root?

rintisch commented:
@extcode  
That's a valid question. I am not sure about it. 
In general I would say yes, the `package.json` and `package-lock.json` be placed in the project root.
But:
* I used https://github.com/jweiland-net/maps2/tree/main/Resources/Private/Build as orientation. There it is solved the way I realized it now.
* These files are not really interesting for the general developer. Therefore it might be a good idea to not have them in the project root to not "pollute" it with even more files.

It would be interesting to see other extensions which needed to solve the same problem. Do you know examples? I could also ask tomorrow in slack, maybe others know the best practice.

In every case: I can move them to the project root if you want, you're the owner.

extcode commented:
A had a look a the TYPO3 Verified Extensions and found different solutions:

- mask had an _Build/Sources_ folder: https://github.com/Gernott/mask/tree/main/Build/Sources
- lux had an _Resources/Private/Build_ folder: https://github.com/in2code-de/lux/tree/develop/Resources/Private/Build

My thoughts on this were only that you can immediately see from the package.json in the root directory that there is JavaScript (or perhaps later also TypeScript) for building and that you can incorporate this into the build pipeline of your own project if necessary.
Probably hardly anyone will build the files themselves, but if you have certain requirements, you could.

But that was just a thought of mine. Otherwise, I don't care where the file is located.

Another question would be whether the package-lock.json is really needed in the project. It would probably be updated every time because there are newer versions. The composer.lock is also left out. But since I'm not an expert, that's more of a question.

rintisch commented:
@extcode 
Thanks for your thoughts.
I made the following adaptions:
* Move source files to `/Build/Sources/` which seems to be the best place because you already has a `/Build` directory.
* `package-lock.json` is not versioned any longer.
* I added so more information in `README.md` files in the source and public directory.
* I added information in the changelog entry.

In the next step I will integrate the JS from EXT:cart_products which was introduced with https://github.com/extcode/cart_products/commit/3249a834e881bf8446c807a46d774c6a12dc3099

extcode commented:
@rintisch I think this is a good solution. You can remove the WIP flag.

rintisch commented:
@extcode I marked the PR as "ready for review". I will make another PR after you merged this to integrate the JS from EXT:cart_products or is there anything against it?

extcode commented:
@rintisch Is there a part of generic JavaScript in extcode/cart-products? Specific JavaScript should stays in the product extension. Lets talk about integration on Slack.
